More detailed info on block scalars (#2319)
ansible/ansible-documentation: docs/docsite/rst/reference_appendices/YAMLSyntax.rst

YAML Basics
-----------

> For Ansible, nearly every YAML file starts with a list.

Reviewer: All play and task files are lists

All files that define variables are dictionaries

Author: I made some comments about this in conversations below, so resolving this thread.

This is part of the YAML format and indicates the start and end of a document.

> One of the most common structures that can be found in a yaml file is a list.

Reviewer 2: ```suggestion
Ansible playbooks and task files are lists.
```

Author: I'd prefer to leave it as it was since at this point we are talking about yaml in general (regardless of its association with Ansible). Also, variables file being a dictionary can still have fields that are lists and thus we can say that lists can be found in those files too even though the files are dictionaries.

Reviewer 3: I agree with @bcoca and @s-hertel in these conversations. While we are talking about yaml in general, this is a yaml helper doc specifically for Ansible so we need to keep the little mentions here and there about Ansible playbooks etc being lists. It helps bring the two concepts together. I suggest using both sentences.

> Each item of a list tarts with a ``"- "`` (a dash and a space) and must be at the same indentation level:

Reviewer 2: ```suggestion
Each item of a list starts with a ``"- "`` (a dash and a space) and must be at the same indentation level:
```

> .. note:: Indentation is the amount of white spaces from the start of a line. Correct indentation is important for yaml documents to be readable by Ansible.

Reviewer 2: ```suggestion
.. note:: Indentation is the amount of whitespace from the start of a line. Correct indentation is important for YAML documents to be valid.
```

    - Mango
    ...

> Another common structure is a dictionary which consists of one or more key/value pairs: ``key: value`` (the colon must be followed by a space):

Reviewer 2: Dictionaries don't need to contain at least one key/value pair.

```suggestion
Ansible variable files are dictionaries.
A YAML dictionary is represented in a simple ``key: value`` form (the colon must be followed by a space):
```

Author: Same as the part about lists. I'd like to keep my version if possible.

Author: > Dictionaries don't need to contain at least one key/value pair.

Meaning empty dictionary? Formally it's correct, of course, but I think for the purposes of the doc it's not relevant really.

Also, if we get into saying that a certain yaml file is a list or a dictionary then the next question would be: is empty yaml file an empty list or empty dictionary, lol? I'm not sure we want to put ourselves in that discussion :-) I'd prefer to keep my version if possible.

Reviewer 3: Agree with @s-hertel. Same idea of keeping the Ansible mentions here. There is nothing in the suggested edits that says anything about empty dictionaries.

          - fortran
          - erlang

> Dictionaries and lists can also be written in a short form called "Flow collections":

Reviewer: aka YAML shorthand, which is also a superset of JSON

Author: Is this a suggested addition?

    likes_emacs: TRUE
    uses_cvs: false

Use lowercase 'true' or 'false' to be compatible with default yamllint options.

> Values can be written as multiple lines using "Block Scalars".

Reviewer: String values

Author: Made suggested change, so resolving.

"Folded Block Scalar" ``>`` will replace carriage returns at the end of each line with spaces. Carriage returns can be added by leaving an empty line.
For example the following:

.. code:: yaml

    fold_newlines: >
                this is really a
                single line of text
                despite appearances

                this is going to be a
                second line of text

> will result in below (note additional new line at the end):

Reviewer 2: ```suggestion
will result in two lines of text followed by one newline:
```

.. code:: text

    this is really a single line of text despite appearances\n
    this is going to be a second line of text\n

> "Literal Block Scalar" ``|`` will keep carriage returns and any trailing spaces (again, note additional new line at the end):

Reviewer 2: ```suggestion
"Literal Block Scalar" ``|`` will keep carriage returns and any trailing spaces (again, followed by one newline):
```
